[RFC] Speculatively disqualify a node which seems to be down or overloaded from read operations

psarna opened the pull request:
This series presents a mechanism of disqualifying a node from read operations if it looks overloaded/down. It's based on estimating how long ago the target node was still responsive - an as that time gets longer than request's timeout, the chance for dropping the request increases.

Things to consider:
 1. Perhaps this mechanism should fire depending on the workload type (#8096, #8097)
 2. While I successfully tested an early version of this solution on AWS, I don't yet have any cool grafana screenshots for this specific version of the source code.
 3. The heuristics for qualifying requests for getting dropped should be subject to debate
 4. I haven't measured the performance penalty for looking up endpoints in a flat map for each sent request, but I strongly suspect it's not going to be noticeable, given that the number of entries is expected to be very small (roughly the number of live nodes in the cluster) and a flat map is used to take advantage of this fact.

/cc @avikivity @gleb-cloudius

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3497,8 +3497,40 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
+            auto last_attempt = now - last_seen->last_sent;
+            // If there was no communication attempt for a long while, the node should be probed in case it came back to life
+            if (last_attempt > time_left) {
+                return false;
```
Is the intent that the caller update last_sent, and so the next call will not enter this path?

Maybe the marking and testing should be combined, so this is explicit. a try-style function.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3497,8 +3497,40 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
+        }
+        if (auto last_seen = _proxy->last_seen(ep)) {
+            auto now = clock_type::now();
+            auto latency_estimate = last_seen->last_sent - last_seen->last_responded;
```
Why is this a latency estimate? These are timestamps for unrelated requests. The difference could be negative, zero, or positive.

psarna replied:
Yeah, the name is unfortunate, it's only a latency estimate in very broad terms with lots of assumption. What I actually want to estimate is "a time period in which the target failed to response", but I don't have any clue how to shorten the name properly.

avikivity replied:
Complicated concepts should have complicated names (or be simplified)

avikivity commented on service/storage_proxy.cc:
```diff
@@ -3515,6 +3547,9 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
         return parallel_for_each(begin, end, [this, resolver = std::move(resolver), timeout] (gms::inet_address ep) {
+            if (maybe_speculatively_fail(*resolver, ep, timeout)) {
+                return make_ready_future<>();
+            }
```
Shouldn't we pick another replica instead?

avikivity replied:
And mark this as an error?

psarna replied:
It's marked as an error by putting the error in the resolver, inside `maybe_speculatively_fail`. Returning an exception_future here would just result in a warning that an exceptional future is ignored, since everything is waited for on the resolver.

As for picking replicas - speculative retry would do that for free if it's enabled and the resolver would not receive enough acks to fulfill the CL. If speculative retry is not enabled, then perhaps we can indeed try being smarter and suggesting another replica, although I expect the patch to be rather intrusive.

gleb-cloudius replied:
So are we re-inventing dynamic snitch?

nyh replied:
I agree with @gleb-cloudius that we should have answered this question before showing the code. I tried to do it in a new issue https://github.com/scylladb/scylla/issues/8964 - I hope @psarna agrees with my description of his idea, and my explanation of why this idea is different (and better) than the dynamic snitch will please @gleb-cloudius.

psarna commented:
v2:
 - renamed `latency_estimate` variable to more fitting `time_without_response`

I'm running some scenarios now, will post results if I see something interesting.

gleb-cloudius commented on service/storage_proxy.cc:
```diff
@@ -3525,6 +3559,9 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
     }
     future<> make_digest_requests(digest_resolver_ptr resolver, targets_iterator begin, targets_iterator end, clock_type::time_point timeout) {
         return parallel_for_each(begin, end, [this, resolver = std::move(resolver), timeout] (gms::inet_address ep) {
+            if (maybe_speculatively_fail(*resolver, ep, timeout)) {
```
Why speculatively fail if we can speculatively choose another replica when we calculate targets?

psarna replied:
We would choose another replica via speculative retry (if configured), which has the advantage that the mechanism already exists and works. The intention is that Scylla would behave exactly as if this target node was already dead/overloaded, but without actually allocating a background read for it, thus saving coordinator's resources.

 Initially I started this series by implementing this mechanism at the point where targets are picked as you suggested, but the change was rather intrusive and I gave it up after coming up with more and more corner cases to consider. Still, perhaps the complexity is worth the effort, since we'd pick another node (if possible) right away, instead of introducing a delay by relying on speculative retry. I'll take another shot at introducing this mechanism sooner, perhaps I changed my mind to soon.

avikivity commented on service/storage_proxy.hh:
```diff
@@ -295,6 +296,11 @@ private:
     std::unordered_map<gms::inet_address, view_update_backlog_timestamped> _view_update_backlogs;
+    struct last_seen_info {
+        clock_type::time_point last_responded = clock_type::time_point::min();
+        clock_type::time_point last_sent = clock_type::time_point::min();
```
e.g. add `bool up` here.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4593,6 +4595,36 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+            static constexpr int granularity = 10'000;
+            static thread_local std::uniform_int_distribution dice(0, granularity);
+            // There's always a small chance of probing the replica in case it came back to life.
+            int fail_chance = std::min<long>((time_without_response.count() - time_left.count()) * granularity / time_left.count(), granularity - 1);
```
Are we certain that time_left is nonzero and nonnegative?

psarna replied:
Hm, as long as timeout > now, which sounds reasonable, but perhaps it's also not guaranteed. That can be short-circuited with a simple check - if we're past timeout already, the node can be safely qualified as dead - as will other nodes, and the query would deservedly fail.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4593,6 +4595,36 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+    }
+    if (auto last_seen_ep = last_seen(ep)) {
+        auto now = clock_type::now();
+        auto time_without_response = last_seen_ep->last_sent - last_seen_ep->last_responded;
```
I'm not sure this is a good metric. Let's say a replica is receiving 1M reqs/sec and responding 1M reqs/sec, but with a latency of 500ms. Is it overloaded or not?

psarna replied:
I'm also not sure, but the rough idea is that users would be able to impact this metric with custom timeouts. Let's say that we have 1M reqs/sec with a latency of 500ms and a default 5s timeout. According to this code, the node is not qualified as overloaded. However, if the user sets an explicit timeout of 200ms for this request, because otherwise he loses a bid, then the node is qualified as overloaded.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4593,6 +4595,36 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+                return false;
+            }
+        }
+    }
```
I like the probabilitic aspect, it gives an overloaded node opportunity to still function.

nyh replied:
I'm not convinced this mechanism can or should detect an "overloaded" node - I think rather it will detect a down or unavailable node. I tried to describe what I **think** this mechanism does and doesn't do in https://github.com/scylladb/scylla/issues/8964 - please tell me if you disagree and there is some useful sense of how this mechanism is meant to detect "overloaded" nodes.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4593,6 +4595,36 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
+                slogger.debug("Overload protection: speculatively disqualifying {} as a candidate for a read request with {}ms left until timeout, "
+                        "since it hasn't responded for at least {}ms", ep, time_left.count(), time_without_response.count());
+                return false;
+            }
```
We need a metric for requests that had a replica disqualified due to speculative overload protection".

avikivity commented on service/storage_proxy.cc:
```diff
@@ -1528,7 +1528,8 @@ storage_proxy_stats::stats::stats()
         , mutation_data_read_completed(COORDINATOR_STATS_CATEGORY, "completed_reads", "number of mutation data read requests that completed", "mutation_data")
-        , mutation_data_read_errors(COORDINATOR_STATS_CATEGORY, "read_errors", "number of mutation data read requests that failed", "mutation_data") { }
+        , mutation_data_read_errors(COORDINATOR_STATS_CATEGORY, "read_errors", "number of mutation data read requests that failed", "mutation_data")
+        , read_requests_speculatively_refused(COORDINATOR_STATS_CATEGORY, "read_requests_speculatively_refused", "number of requests which were destined for this endpoint, but speculatively refused", "data") { }
```
Just as I predicted.

avikivity commented:
I'd like to see justification for the calculation. I don't think the code is a good place for it, a short paper please.

psarna commented:
Sure, I'm also in the process of producing colorful grafana screenshots, which would hopefully help me defend this case. I'll also add some rationale to docs/.

psarna commented:
Here's how read request metrics look like with the new code with the following scenario:
1. A 3-node cluster is preloaded with some data, rf=3, cl=all
2. A quorum read workload starts - with custom `USING TIMEOUT 1s` CQL predicate
3. One node is killed during the workload

The point when the node was killed can be identified by a huge red line and the fact that the read rate for the yellow node starts dropping and the number of refusals starts rising. Notice that it takes ~30 seconds for the coordinator to acknowledge the killed node a down (it can also be identified by looking at the bottom graph and seeing where the yellow flat line stops), which means that it would try to contact this node during this whole period and produce fruitless background reads. Instead, the node is quickly qualified as overloaded/dead and the rate of requests sent to it decreases.

The yellow node is later restarted, which can be identified as the increase of its read rate.

![speculatively_refused](https://user-images.githubusercontent.com/10433047/113131502-98c9c900-921d-11eb-88c1-662f0c0bf2a2.png)

psarna commented:
v4:
 - added a short-circuit refusal when the request is already past its timeout - no point in estimating if the replica is able to respond in a negative period of time
 - added a short doc explaining the idea

avikivity commented:
> Here's how read request metrics look like with the new code with the following scenario:
> 
> 1. A 3-node cluster is preloaded with some data, rf=3, cl=all
> 2. A quorum read workload starts - with custom `USING TIMEOUT 1s` CQL predicate
> 3. One node is killed during the workload
> 
> The point when the node was killed can be identified by a huge red line and the fact that the read rate for the yellow node starts dropping and the number of refusals starts rising. Notice that it takes ~30 seconds for the coordinator to acknowledge the killed node a down (it can also be identified by looking at the bottom graph and seeing where the yellow flat line stops), which means that it would try to contact this node during this whole period and produce fruitless background reads. Instead, the node is quickly qualified as overloaded/dead and the rate of requests sent to it decreases.
> 
> The yellow node is later restarted, which can be identified as the increase of its read rate.
> 
> ![speculatively_refused](https://user-images.githubusercontent.com/10433047/113131502-98c9c900-921d-11eb-88c1-662f0c0bf2a2.png)

This is quite nice. Note publishing nice graphs puts you at risk for being volunteered into writing a blog post.

psarna commented:
Tomorrow I'll spin an aws cluster and try to inspect the metrics for a 3-node-cluster-with-one-node-being-slightly-slower scenario. Hopefully I could observe the slower node getting less traffic with sufficiently large workload, with some queries getting redirected to other nodes.

psarna commented:
Alright, yesterday's grafana output lacked context and it turns out to not be very accurate - the drop of read requests can also be observed before the patch, because this particular metrics represents completed reads - which are naturally dropping for a dead node anyway.

Here's a more useful comparison - the same scenario is ran first without the patch, and then with it:

1. A 3-node cluster is preloaded with some data, rf=3, cl=all
2. A quorum read workload starts - with custom USING TIMEOUT 1s CQL predicate; cl=quorum
3. One node is killed during the workload

The red line in the middle indicates the point in which the cluster was upgraded to a version with the patches applied.

What can be observed is that in the first case, right after one of the replicas is killed, we can observe a ~30s window of sharp increase in background reads along with accompanying read errors, as well as elevated number of digest reads, which is a result of speculative retry firing off.

In the second case, background reads + read errors + digest reads are quickly translated to the new metrics - refused reads, which indicates that the node was disqualified as a candidate and the requests were rerouted to other nodes instead.

![with_and_without](https://user-images.githubusercontent.com/10433047/113267191-587a5180-92d6-11eb-85f4-f081dab0dbac.png)

avikivity commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,69 @@
+ * each time a read request is sent to a replica, a timestamp is recorded
+ * each time a response is received from a replica, a timestamp is recorded
+ * the difference between these two timestamps is a cheap, rough estimate
+   of how long it may take this replica to respond
```
It's not. The difference can be positive, negative, or zero, regardless of how overloaded the replica is.

avikivity commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,69 @@
+a `sent` timestamp and a `responded` timestamp belong to the same request,
+so these values may vary. However, assuming that the cluster is under
+continuous, uniform load, these values would get updated very often, and would
+hopefully approach the approximate round-trip time for a single request.
```
Why would they? For a heavily used cluster, we'd be continuously sending requests and receiving responses. A replica could be responding 4 seconds too late, and still have continuous requests and responses.

psarna replied:
Hm, right. I still haven't run an aws-based workload with a slower node, but that's what I'd most likely discover.

avikivity commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,69 @@
+so these values may vary. However, assuming that the cluster is under
+continuous, uniform load, these values would get updated very often, and would
+hopefully approach the approximate round-trip time for a single request.
+Finally, when a replica was not contacted in a long while (e.g. 5s),
```
I think this is the part that works.

Maybe we could estimate overload by looking at response time for past requests. So if we had a QUORUM read and replica A responded in 1ms and replica B responded in 5ms, that gives us some information about A and B. But it doesn't work for LOCAL_ONE queries.

psarna replied:
Yup, it would make more sense. It's actually also trivial to gather the response time for a single request. I'll take it into consideration for the next iteration, to actually make this estimate work in cases other than a killed node - in particular, a replica which constantly replies, but with a large lag.

avikivity replied:
I think we should brainstorm this more. Maybe look at literature too.

psarna replied:
Certainly. I'll add an RFC prefix to this series to indicate that it's not really ready yet

avikivity commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,69 @@
+Based on the last seen information, the coordinator is now allowed
+to make an educated guess whether a replica is going to respond in time.
+A simple, probabilistic approach is implemented as follows.
+At first, we compute how much time left (`TL`) the request still has until
```
This number is also dangerous. Usually it's similar to the timeout. But if the coordinator is overloaded, it gets small (or even zero or negative). If it appears in the denominator of a division, it gets amplified.

psarna replied:
Right. In the denominator, it would perhaps make sense to use the timeout duration, which is less prone to getting really small (unless explicitly specified by the user). That one would need to be passed everywhere explicitly, but it's lightweight.

avikivity replied:
It also feels fishy. It makes a workload that has low timeouts be very sensitive to extra delays. These delays could ripple (a problem with dynamic snitch): a node gets a little slow, other cooridnators back off and send requests to another node; that node becomes slow (and the original node is now fast) and traffic flip-flops between the two.

I think the failure detector part in this patch works and was tested, and the overload backoff wasn't.

psarna commented:
I pushed a new version in which the round-trip time for requests is gathered in order to handle a case with an overloaded node which responds regularly, but the time spent on a single request still reaches/exceeds the timeout. I'll put this version to more tests, and meanwhile also try to find some points of reference how similar issues are solved.

I still haven't addressed the problem of an overloaded coordinator, when the time left for a request to be handled is already very short by the time we decide which replicas to pick as targets. Actually, perhaps another mechanism should be in place - if we determine that there's not enough time to reliably send requests to replicas, and it's mainly because the request was severely delayed on the coordinator side - the coordinator should instead return OverloadedException, which is why it was created in the first place. Of course, question remains how exactly to decide if the coordinator is overloaded or not.

psarna commented:
It's truly amazing how tackling a single overload-related issue usually results in 1 partial solution and at least 2 more fully-fledged overload-related issues. It's a fork bomb.

psarna commented on docs/design-notes/qualifying-replicas-for-reads.md:
```diff
@@ -0,0 +1,79 @@
+the previous request and the current request is going to time out in X milliseconds,
+it gets more and more likely to be dead.
+The last recorded round trip time is, in turn, a cheap, rough estimate of how long
+it may take the replica to respond to the next request - assuming that the conditions
```
Note to self: also bad, because a momentary spike in latency can disqualify a replica for a couple of seconds. Something more resistant to spikes is needed (moving average?)

nyh replied:
In https://github.com/scylladb/scylla/issues/8964 I tried to explain how what this mechanism is supposed to measure (the duration in which we got **no response**) is not actually related to latency, but rather to throughput - even when latencies are getting high, you don't normally have a very long duration of getting no responses at all.
Please tell me if I'm misunderstanding something....

psarna commented:
@avikivity I gave it some thought and I think that it makes sense to split this problem into two:
 1. better failure detection for replicas during reads
 2. overload detection for replicas during reads

... and make this series only address (1.). The reason is that this series starts having symptoms of lingering in review for months, because (2.) is hard and requires research. (1.) is quite simple however and once https://github.com/scylladb/scylla/issues/3746 is fixed, the problem of sharp increase in background reads when a node is killed will be more severe, because https://github.com/scylladb/scylla/issues/3746 currently acts as backpressure for it.
And all it actually takes for this series to fully address (1.) is to stop claiming in docs and description that it also addresses (2.), which it doesn't. What do you think?

psarna commented:
(also, this pull request suffered from too many pivots already, so I'll open a new one with failure detection only)

psarna commented:
not going to be merged, closing
